### PR TITLE
feat: ignore_delete_errors variable

### PR DESCRIPTION
**Pull request by luke-zhu**
Currently, if the Python code returns an exception, the user needs to delete the CloudFormation Stack through the UI or CLI. This change makes it so that a variable can be changed to ignore deletion issues.

Specifically, this addresses the case where log_group_matches is invalid regex and the user wants to change it.

The variable is disabled by default since enabling by default means that we might leave subscription filters behind

### CODE REVIEW

File: lambda/index.py [hunk 1]
```diff
@@ -337,6 +345,9 @@ def main(event, context):
     delivery_role = os.environ['DELIVERY_STREAM_ROLE_ARN']
     timeout = os.environ['TIMEOUT']
 
+    global ignore_delete_errors
+    ignore_delete_errors = os.environ['IGNORE_DELETE_ERRORS']
```
Comment (jta) [inline lambda/index.py:349]: Set the environment variable directly when declaring `ignore_delete_errors` (line 32).

File: variables.tf [hunk 2]
```diff
@@ -94,6 +94,16 @@ variable "lambda_memory" {
...
 }
 
+variable "ignore_delete_errors" {
+  description = <<-EOF
+    Ignore CloudFormation stack errors from deletion events.
+
+    Setting this to true means that leftover Subscription Filters could remain.
+  EOF
+  type        = bool
+  default     = false
```
Comment (jta) [inline variables.tf:104]: Would `true` be a better default? Seems odd that you'd have to stall to find out what this option does.

Comment (luke-zhu) [reply variables.tf:104]: I think it's a more usable default. Users just might be surprised by it if we don't clean up properly.

File: variables.tf [hunk 3]
```diff
@@ -94,6 +94,16 @@ variable "lambda_memory" {
   default     = 128
 }
 
+variable "ignore_delete_errors" {
+  description = <<-EOF
+    Ignore CloudFormation stack errors from deletion events.
+
+    Setting this to true means that leftover Subscription Filters could remain.
```
Comment (bendrucker) [inline variables.tf:101]:
```suggestion
  description = "If an error occurs while deleting subscription filters, ignore it, leaving behind any remaining filters."
```

**Review by bendrucker — COMMENTED** [review]
Leaving a bunch of subscription filters behind so that the user can delete a CloudFormation stack whose only job is to invoke a Lambda doesn't seem to really fix anything. Would users actually be ok leaving all the subscription filters around? Are other Observe resources (e.g. Kinesis firehose) prevented from deletion because they are still targeted by subscription filters?

You are very limited in the number of filters you can attach, so orphaning them is consequential:

https://aws.amazon.com/about-aws/whats-new/2020/10/amazon-cloudwatch-logs-now-supports-two-subscription-filters-per-log-group/

I need some clarification on the nature of these errors and the resulting states to understand whether ignoring them is truly the only option. My default expectation would be that when the CloudFormation custom resource deletion fails, the overall stack deletion fails, and thus `terraform destroy` fails. The user can then retry as needed, after suffering through the latency of a CloudFormation run. 

The original description suggests that this is not the case you're after, but I don't have enough to understand what's going on. Examples of the sequence of Terraform configurations that can be applied to reproduce this and the output that results would be a good start.

File: variables.tf (hunk 3 above)
Comment (luke-zhu) [reply variables.tf:111]: changed

File: README.md [hunk 4]
```diff
@@ -88,6 +88,7 @@ No modules.
 | <a name="input_filter_pattern"></a> [filter\_pattern](#input\_filter\_pattern) | The filter pattern to use. For more information, see [Filter and Pattern Syntax](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/FilterAndPatternSyntax.html)" | `string` | `""` | no |
 | <a name="input_iam_name_prefix"></a> [iam\_name\_prefix](#input\_iam\_name\_prefix) | Prefix used for all created IAM roles and policies | `string` | `"observe-logs-subscription"` | no |
 | <a name="input_iam_role_arn"></a> [iam\_role\_arn](#input\_iam\_role\_arn) | ARN of IAM role to use for Cloudwatch Logs subscription.<br>If this is not specified, then an IAM role is created. | `string` | `""` | no |
+| <a name="input_ignore_delete_errors"></a> [ignore\_delete\_errors](#input\_ignore\_delete\_errors) | Ignore CloudFormation stack errors from deletion events.<br><br>Setting this to true means that leftover Subscription Filters could remain. | `bool` | `false` | no |
```
Comment (bendrucker) [inline README.md:91]: docs generator needs to be re-run to update the description here

**Review by bendrucker — COMMENTED** [review]
Will approve this when the description is updated, but I still think this deserves more documentation. This doesn't necessarily have to be in the module (i.e. variable descriptions) and might deserve a separate readme section. E.g., if we know that if you have >1,000 log groups you are likely to see failures and should first try increasing the timeout and then maybe ignore the errors, we should clearly state that with examples. As is this seems like something the customer would need to debug themselves or have us point them to. If it's a known failure mode, we can do better.

File: README.md (hunk 4 above)
Comment (luke-zhu) [reply README.md:91]: done

**Comment by luke-zhu** [comment]
I added a troubleshooting section to the README and updated the cloudformation too

**Review by bendrucker — APPROVED** [review]

**Review by chenjiayuan — APPROVED** [review]

**Pull request merged**